File: lintcode/python/0760_binary_tree_right_side_view.py
#!/usr/bin/python -t

# bfs

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

from collections import deque

class Solution:
    """
    @param root: the root of the given tree
    @return: the values of the nodes you can see ordered from top to bottom
    """
    def rightSideView(self, root):
        # write your code here
        ret = []
        if not root:
            return ret
            
        q = deque([root])
        ret.append(root.val)
        
        while len(q) > 0:
            l = len(q)
            
            level = []
            for i in range(l):
                node = q.popleft()
            
                if node.left:
                    level.append(node.left.val)
                    q.append(node.left)
                    
                if node.right:
                    level.append(node.right.val)
                    q.append(node.right)
            if level:
                ret.append(level[-1])
            
        return ret
        

# dfs

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

class Solution:
    """
    @param root: the root of the given tree
    @return: the values of the nodes you can see ordered from top to bottom
    """
    def rightSideView(self, root):
        # write your code here
        if not root:
            return []
            
        self.ret = []
        
        self.dfs(root, 0)
        
        return self.ret
        
    def dfs(self, root, depth):
        if not root:
            return
        if depth == len(self.ret):
            self.ret.append(root.val)

        self.dfs(root.right, depth+1)
        self.dfs(root.left, depth+1)
            
        return
    
    

File: lintcode/golang/0760_binary_tree_right_side_view.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

/**
 * @param root: the root of the given tree
 * @return: the values of the nodes you can see ordered from top to bottom
 */
func rightSideView (root *TreeNode) []int {
    // write your code here
    ret := []int{}
    if root == nil {
        return ret
    }
    
    q := make([]*TreeNode, 0)
    q = append(q, root)
    
    for len(q) != 0 {
        level := make([]*TreeNode, 0)
        node := q[len(q)-1]
        ret = append(ret, node.Val)
        for _, n := range q {
            if n.Left != nil {
                level = append(level, n.Left)
            }
            if n.Right != nil {
                level = append(level, n.Right)
            }
        }
        q = level
    }
    
    return ret
}
